crates/ty_python_semantic/resources/mdtest/generics/pep695/paramspec.md

```diff
@@ -62,3 +62,503 @@ Other values are invalid.
+    def wrapper(*args: P.args, **kwargs: P.kwargs) -> bool:
+        func1(*args, **kwargs)
+        func2(*args, **kwargs)
+        return True
+    return wrapper
+```
+
+As per the spec,
+
+> A user may include the same `ParamSpec` multiple times in the arguments of the same function, to
+> indicate a dependency between multiple arguments. In these cases a type checker may choose to
+> solve to a common behavioral supertype (i.e. a set of parameters for which all of the valid calls
+> are valid in both of the subtypes), but is not obligated to do so.
+
+TODO: Currently, we don't do this
+
+```py
+def xy(x: int, y: str) -> int:
+    return 1
+
+def yx(y: int, x: str) -> int:
+    return 2
+
+reveal_type(multiple(xy, xy))  # revealed: (x: int, y: str) -> bool
+
+# The common supertype is `(int, str, /)` which is converting the positional-or-keyword parameters
+# into positional-only parameters because the position of the types are the same.
+# TODO: This shouldn't error
+# error: [invalid-argument-type]
+reveal_type(multiple(xy, yx))  # revealed: (x: int, y: str) -> bool
+
+def keyword_only1(*, x: int) -> int:
+    return 1
+
+def keyword_only2(*, y: int) -> int:
+    return 2
+
+# On the other hand, combining two functions with only keyword-only parameters does not have a
+# common supertype, so it should result in an error.
+# error: [invalid-argument-type] "Argument to function `multiple` is incorrect: Expected `(*, x: int) -> int`, found `def keyword_only2(*, y: int) -> int`"
+reveal_type(multiple(keyword_only1, keyword_only2))  # revealed: (*, x: int) -> bool
+```
+
+### Constructors of user-defined generic class on `ParamSpec`
+
+```py
+from typing import Callable
+
+class C[**P]:
+    f: Callable[P, int]
+
+    def __init__(self, f: Callable[P, int]) -> None:
+        self.f = f
+
+def f(x: int, y: str) -> bool:
+    return True
+
+c = C(f)
+reveal_type(c.f)  # revealed: (x: int, y: str) -> int
+```
+
+### `ParamSpec` in prepended positional parameters
+
+> If one of these prepended positional parameters contains a free `ParamSpec`, we consider that
+> variable in scope for the purposes of extracting the components of that `ParamSpec`.
+
+```py
+from typing import Callable
+
+def foo1[**P1](func: Callable[P1, int], *args: P1.args, **kwargs: P1.kwargs) -> int:
+    return func(*args, **kwargs)
+
+def foo1_with_extra_arg[**P1](func: Callable[P1, int], extra: str, *args: P1.args, **kwargs: P1.kwargs) -> int:
+    return func(*args, **kwargs)
+
+def foo2[**P2](func: Callable[P2, int], *args: P2.args, **kwargs: P2.kwargs) -> None:
+    foo1(func, *args, **kwargs)
+
+    # error: [invalid-argument-type] "Argument to function `foo1` is incorrect: Expected `P2@foo2.args`, found `Literal[1]`"
+    foo1(func, 1, *args, **kwargs)
+
+    # error: [invalid-argument-type] "Argument to function `foo1_with_extra_arg` is incorrect: Expected `str`, found `P2@foo2.args`"
+    foo1_with_extra_arg(func, *args, **kwargs)
```

Comment:
and also add a call to `foo1_with_extra_arg` that _doesn't_ cause us to emit a diagnostic?

```suggestion
    # error: [invalid-argument-type] "Argument to function `foo1_with_extra_arg` is incorrect: Expected `str`, found `P2@foo2.args`"
    foo1_with_extra_arg(func, *args, **kwargs)
    
    foo1_with_extra_arg(func, "foooooo", *args, **kwargs)
```